### PR TITLE
feat(datasource/pypi): always fallback to try simple endpoints

### LINKED ISSUE #26483: PyPI datasource fall back to simple API if JSON API fails
### Describe the proposed change(s).

Similar to how we query rubygems.org, we should use the simple API to get the list / check for changes and then secondarily see if the rich API exists for querying metadata. 

Ie instead of falling back to simple as we do today, we should fall back to API instead, or enrich using it 

Algorithm:
- Query simple API first
- If simple fully fails, try the JSON API and return results
- If simple succeeds, check JSON API cache to see if we have metadata for all releases. If we do, no need to query the JSON API and we can return the combined results
- If the JSON API cache does not have metadata for all, query it, cache the new results, and return the merged data
- If JSON API fails, but simple worked, return simple results

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Modifies pypi datasource so that it _always_ attempts to fall back to a simple result if the JSON API throws an error.

## Context

Built on top of #26386 (thanks @Kakadus) and #31771 (thanks @samgiz).

Closes #26483
Closes #26383

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

Closes #26483
Closes #26383

### CODE REVIEW

**Comment by samgiz** [comment]
Do you need any help from me to get this merged?

File: lib/modules/datasource/pypi/index.ts [hunk 1]
```diff
@@ -66,10 +66,6 @@ export class PypiDatasource extends Datasource {
         // we need to resolve early here so we can catch any 404s and fallback to a simple lookup
         dependency = await this.getDependency(normalizedLookupName, hostUrl);
       } catch (err) {
-        if (err.statusCode !== 404) {
-          throw err;
-        }
-
         // error contacting json-style api -- attempt to fallback to a simple-style api
         logger.trace(
           { packageName, hostUrl },
```
Comment (viceice) [inline lib/modules/datasource/pypi/index.ts:71]: we should log the error on trace for troubleshooting

**Comment by samgiz** [comment]
Is anything stopping this from being merged?

**Review by viceice — APPROVED** [review]

**Review by Kakadus — APPROVED** [review]
nice to see this getting addressed after some time :tada: 

Sadly, I were not able to proceed my initial PR as priorities changed and time ran short. Thank you for yours and for renovate!

**Pull request merged**